File: 필수문항/138_용돈관리.py
import sys
import math

input = sys.stdin.readline

N, M = map(int, input().split())
moneys = [int(input()) for _ in range(N)]
sum_moneys = sum(moneys)
lp = max(max(moneys), math.ceil(sum_moneys / M))
rp = sum_moneys

while lp <= rp:
    mid = (lp + rp) // 2
    tmp_sum = 0
    tmp_ans = 1
    for money in moneys:
        if tmp_sum + money > mid:
            tmp_ans += 1
            tmp_sum = money
        else:
            tmp_sum += money
    if tmp_ans > M:
        lp = mid + 1
    else:
        rp = mid - 1

print(lp)

File: 필수문항/138_용돈관리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int[] moneys = new int[N];
        int max = 0;
        int sum = 0;
        for (int i = 0; i < N; i++) {
            moneys[i] = Integer.parseInt(br.readLine());
            max = Math.max(max, moneys[i]);
            sum += moneys[i];
        }
        int lp = max;
        int rp = sum;
        while (lp <= rp) {
            int mid = (lp + rp) / 2;
            int tmpAns = 1;
            int tmpSum = 0;
            for (int money : moneys) {
                if (money + tmpSum > mid) {
                    tmpAns++;
                    tmpSum = money;
                } else {
                    tmpSum += money;
                }
            }
            if (tmpAns > M) {
                lp = mid + 1;
            } else {
                rp = mid - 1;
            }
        }
        System.out.println(lp);
    }
}
